wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/devfile/OverridePropertiesApplier.java

```diff
@@ -0,0 +1,114 @@
+    for (Map.Entry<String, String> entry : overrideProperties.entrySet()) {
+      String[] pathSegments = entry.getKey().split("\\.");
+      if (pathSegments.length < 1) {
+        continue;
+      }
+      validateFirstSegment(pathSegments);
+      String lastSegment = pathSegments[pathSegments.length - 1];
+      JsonNode currentNode = devfileNode;
+      Iterator<String> pathSegmentsIterator = Stream.of(pathSegments).iterator();
+      do {
+        String currentSegment = pathSegmentsIterator.next();
+        JsonNode nextNode = currentNode.path(currentSegment);
+        if (nextNode.isMissingNode() && pathSegmentsIterator.hasNext()) {
+          // no such intermediate node, let's create it as a empty object
+          currentNode = ((ObjectNode) currentNode).putObject(currentSegment);
+          continue;
+        } else if (nextNode.isArray()) {
+          // ok we have reference to array, so need to make sure that we have next path segment
+          // ant then try to retrieve it from array
+          if (!pathSegmentsIterator.hasNext()) {
+            throw new DevfileFormatException(
+                format(
+                    "Override property reference '%s' points to an array type object. Please add an item qualifier by name.",
+                    entry.getKey()));
+          }
+          // retrieve object by name from array
+          String arrayObjectName = pathSegmentsIterator.next();
+          currentNode = findNodeByName((ArrayNode) nextNode, arrayObjectName, currentSegment);
+          continue;
+        } else {
+          // because it's impossible to change value of the current Json node,
+          // so to set value, we should be 1 level up and do put(key, value),
+          // so not set latest segment as an current node
+          if (pathSegmentsIterator.hasNext()) {
+            currentNode = nextNode;
+          }
+        }
+      } while (pathSegmentsIterator.hasNext());
+      // end of path segments reached, now we can set value
+      ((ObjectNode) currentNode).put(lastSegment, entry.getValue());
+    }
+    return devfileNode;
+  }
+
+  private void validateFirstSegment(String[] pathSegments) throws DevfileFormatException {
+    if (!allowedFirstSegments.contains(pathSegments[0])) {
+      throw new DevfileFormatException(
+          format(
+              "Override path '%s' starts with an unsupported field pointer. Supported fields are %s.",
+              join(".", pathSegments),
+              allowedFirstSegments.stream().collect(joining("\",\"", "{\"", "\"}"))));
+    }
+  }
+
+  private JsonNode findNodeByName(ArrayNode parentNode, String searchName, String parentNodeName)
+      throws DevfileFormatException {
+    return StreamSupport.stream(parentNode.spliterator(), false)
+        .filter(node -> node.path("name").asText().equals(searchName))
+        .findFirst()
+        .orElseThrow(
+            () ->
+                new DevfileFormatException(
```

Comment:
Not sure if `DevfileFormatException` is the right one here - there's nothing wrong with the devfile itself - we only use this method when trying to apply overrides.